packages/hdwallet-portis/src/portis.ts

```diff
@@ -0,0 +1,386 @@
+  }
+
+  public clearSession(): Promise<void> {
+    return this.portis.logout()
+  }
+
+  public ping(msg: Ping): Promise<Pong> {
+    // no ping function for Portis, so just returning Pong
+    return Promise.resolve({ msg: msg.msg })
+  }
+
+  public sendPin(pin: string): Promise<void> {
+    // no concept of pin in Portis
+    return Promise.resolve()
+  }
+
+  public sendPassphrase(passphrase: string): Promise<void> {
+    // cannot send passphrase to Portis. Could show the widget?
+    return Promise.resolve()
+  }
+
+  public sendCharacter(charater: string): Promise<void> {
+    // no concept of sendCharacter in Portis
+    return Promise.resolve()
+  }
+
+  public sendWord(word: string): Promise<void> {
+    // no concept of sendWord in Portis
+    return Promise.resolve()
+  }
+
+  public cancel(): Promise<void> {
+    // no concept of cancel in Portis
+    return Promise.resolve()
+  }
+
+  public wipe(): Promise<void> {
+    return Promise.resolve()
+  }
+
+  public reset(msg: ResetDevice): Promise<void> {
+    return Promise.resolve()
+  }
+
+  public recover(msg: RecoverDevice): Promise<void> {
+    // no concept of recover in Portis
+    return Promise.resolve()
+  }
+
+  public loadDevice (msg: LoadDevice): Promise<void> {
+    return this.portis.importWallet(msg.mnemonic)
+  }
+
+  public async ethSupportsNetwork (chainId: number = 1): Promise<boolean> {
+    return this.info.ethSupportsNetwork(chainId)
+  }
+
+  public async ethSupportsSecureTransfer (): Promise<boolean> {
+    return this.info.ethSupportsSecureTransfer()
+  }
+
+  public async ethSupportsNativeShapeShift (): Promise<boolean> {
+    return this.info.ethSupportsNativeShapeShift()
+  }
+
+  public async ethVerifyMessage (msg: ETHVerifyMessage): Promise<boolean> {
+    const signingAddress = await this.web3.eth.accounts.recover(msg.message, ('0x' + msg.signature), false)
+    return signingAddress === msg.address
+  }
+
+  public describePath (msg: DescribePath): PathDescription {
+    switch (msg.coin) {
+    case 'Ethereum':
+      return describeETHPath(msg.path)
+    default:
+      throw new Error("Unsupported path")
+    }
+  }
+
+  public ethNextAccountPath (msg: ETHAccountPath): ETHAccountPath | undefined {
+    // Portis only supports one account for eth
+    return undefined
+  }
+
+  public async isInitialized (): Promise<boolean> {
+    return false
+  }
+
+  public disconnect (): Promise<void> {
+    return Promise.resolve()
+  }
+
+  public async getPublicKeys(msg: GetPublicKey[]): Promise<PublicKey[]> {
+    const publicKeys = []
+    this.portisCallInProgress = new Promise( async (resolve, reject) => {
+      await this.portisCallInProgress
+      for (let i = 0; i < msg.length; i++) {
+        const { addressNList } = msg[i];
+        const portisResult = await this.portis.getExtendedPublicKey(addressNListToBIP32(addressNList))
+        const { result, error } = portisResult
+        if(error) {
+          console.log('getPublicKeys error', error)
```

Comment:
should throw something here. the caller assumes the input list, and output list have the same length, and will be misled / confused by this.